src/util/bson/bson.hpp

```diff
@@ -32,41 +32,670 @@
+                return int32_val == other.int32_val;
+            case INT64:
+                return int64_val == other.int64_val;
+            case BOOL:
+                return bool_val == other.bool_val;
+            case DOUBLE:
+                return double_val == other.double_val;
+            case TIME:
+                return time_val == other.time_val;
+            case DATE:
+                return date_val == other.date_val;
+            case OID:
+                return oid_val == other.oid_val;
+            case DECIMAL:
+                return decimal_val == other.decimal_val;
+            case MAX_KEY:
+                return max_key_val == other.max_key_val;
+            case MIN_KEY:
+                return min_key_val == other.min_key_val;
+            case STRING:
+                return *string_val == *other.string_val;
+            case REGEX:
+                return *regex_val == *other.regex_val;
+            case BINARY:
+                return *binary_val == *other.binary_val;
+            case DOCUMENT:
+                return *document_val == *other.document_val;
+            case ARRAY:
+                return *array_val == *other.array_val;
+        }
+    }
+
+    bool operator!=(const Bson& other) const
+    {
+        return !(*this == other);
+    }
+
+private:
+    friend std::ostream& operator<<(std::ostream& out, const Bson& m);
+    template <typename T>
+    friend bool holds_alternative(const Bson& bson);
+
+    typedef enum Type {
+        NONE,
+        INT32,
+        INT64,
+        BOOL,
+        DOUBLE,
+        STRING,
+        BINARY,
+        TIME,
+        DATE,
+        OID,
+        DECIMAL,
+        REGEX,
+        MAX_KEY,
+        MIN_KEY,
+        DOCUMENT,
+        ARRAY
+    } Type;
+    Type m_type;
+
+    union {
+        int32_t int32_val;
+        int64_t int64_val;
+        bool bool_val;
+        double double_val;
+        time_t time_val;
+        Timestamp date_val;
+        ObjectId oid_val;
+        Decimal128 decimal_val;
+        MaxKey max_key_val;
+        MinKey min_key_val;
+        // ref types
+        RegularExpression* regex_val;
+        std::string* string_val;
+        std::vector<char>* binary_val;
+        IndexedMap<Bson>* document_val;
+        std::vector<Bson>* array_val;
+    };
 };
-} // namespace _impl
 
-using Bson = _impl::Fix<_impl::Var>;
+inline Bson::Bson(int32_t v) noexcept
+{
+    m_type = Bson::INT32;
+    int32_val = v;
+}
+
+inline Bson::Bson(int64_t v) noexcept
+{
+    m_type = Bson::INT64;
+    int64_val = v;
+}
+
+inline Bson::Bson(bool v) noexcept
+{
+    m_type = Bson::BOOL;
+    bool_val = v;
+}
+
+inline Bson::Bson(double v) noexcept
+{
+    m_type = Bson::DOUBLE;
+    double_val = v;
+}
+
+inline Bson::Bson(MinKey v) noexcept
+{
+    m_type = Bson::MIN_KEY;
+    min_key_val = v;
+}
+
+inline Bson::Bson(MaxKey v) noexcept
+{
+    m_type = Bson::MAX_KEY;
+    max_key_val = v;
+}
+inline Bson::Bson(const RegularExpression& v) noexcept
```

Comment:
These constructors which allocate memory aren't `noexcept` as the allocation can throw.